### DISCUSSION TITLE
Cookie is not being set

### DISCUSSION THREAD

**Discussion by TomS-** [directus/directus#9720] [Archive]
Since updating the 9.0.0 I'm having issues with login persisting. For the login function I am doing:
```
await directus.auth.login(
            { email: email, password: password },
            { refresh: { auto: true } }
);
```
However, I get the following error after some time:
```
tslib.es6.js:74 Uncaught (in promise) Error: Token expired.
    at Transport.<anonymous> (index.js:1)
    at Generator.throw (<anonymous>)
    at rejected (tslib.es6.js:72)
```
In previous versions I would be kept logged in and the token would auto refresh as expected. 

### To Reproduce

Use the Javascript SDK and use `{ refresh: { auto: true } }` the token will not auto refresh. 

### Errors Shown

```
tslib.es6.js:74 Uncaught (in promise) Error: Token expired.
    at Transport.<anonymous> (index.js:1)
    at Generator.throw (<anonymous>)
    at rejected (tslib.es6.js:72)
```

**Comment by joselcvarela** [comment]
Hello @TomS- 
In fact our SDK was refactored. You can see what have changed here: https://github.com/directus/directus/pull/9080
But this could still be an issue. In order to triage, I need you to post how are you setting up the SDK, making the login and how are you fetching the items.

**Comment by joselcvarela** [comment]
Also, what value has your `ACCESS_TOKEN_TTL`?

**Comment by TomS-** [comment]
Hi @joselcvarela 

I'm setting up the SDK by:
```
import { Directus } from '@directus/sdk';
const directus = new Directus('https://portal.***.dev/');
```
I am then doing:
```
await directus.auth.login({
            email: email,
            password: password 
});
```
(Updated based on your link, auto-refresh is on by default, but the problem still persists)

I am fetching items by:
```
const printJobsItems = await directus.items('print_jobs').readMany({
            filter: {
                        id: { _in: salesOrder.print_production }
            }
});
```

My `.env` settings for access tokens are as follows:
```
ACCESS_TOKEN_TTL="15m"
REFRESH_TOKEN_TTL="7d"
REFRESH_TOKEN_COOKIE_SECURE=false
REFRESH_TOKEN_COOKIE_SAME_SITE="lax"
REFRESH_TOKEN_COOKIE_NAME="directus_refresh_token"
```

**Comment by joselcvarela** [comment]
Thank you @TomS- 
After the login, can you tell how much time has passed until the error is thrown?
My tests were similar to your configuration and didn't have issues.

Also, can you make the same requests using the API endpoints?
It could also be some misconfiguration or bug in API side.

**Comment by TomS-** [comment]
Hi @joselcvarela

It does it exactly on 15 minutes. From what I can tell, if I inspect element and go to `Application > Cookies` there is no `refresh_token` set.

EDIT:
![image](https://user-images.githubusercontent.com/4595662/141102616-51c78e3b-2a72-4682-90a9-a38950acd306.png)
LocalStorage is working, but Cookies are not being set:
![image](https://user-images.githubusercontent.com/4595662/141102717-12bfaa2b-06bb-4da9-8bdc-404813798a59.png)

Could server configuration effect this?

**Comment by joselcvarela** [comment]
It seems to be something with your server because the SDK or any JS side has access to cookies.
Actually I am currently testing this. I already have test Nodejs and no problem found. And, right now, I am keeping a tab open making requests and also no problem found.

So, since you are no cookie set I believe there's some misconfiguration.
Can you check if `login` request has `Set-Cookie` on response headers? In fact, it would be useful to post a screenshot here. Maybe the domain on cookie could be misconfigured.

**Comment by joselcvarela** [comment]
Just found an interesting scenario but it is totally expected.
So I was running SDK tests under Profile A on browser.
Also, I have opened the `/admin` under the same profile.

This can leads to unexpected behaviours like cookie not be present, because both will share the cookie.
Do you think you have done something similar?
If so, try with 
```js
const sdk = new Directus("http://localhost:8055", { auth: { mode: 'json' } });
```

Although this is not advised on browsers. Check here why: https://docs.directus.io/reference/sdk/#options.auth

___
Testing on anonymous tab seems to work as expected, since you not login into admin.

**Comment by TomS-** [comment]
> It seems to be something with your server because the SDK or any JS side has access to cookies. Actually I am currently testing this. I already have test Nodejs and no problem found. And, right now, I am keeping a tab open making requests and also no problem found.
> 
> So, since you are no cookie set I believe there's some misconfiguration. Can you check if `login` request has `Set-Cookie` on response headers? In fact, it would be useful to post a screenshot here. Maybe the domain on cookie could be misconfigured.

![image](https://user-images.githubusercontent.com/4595662/141275304-72ac1da5-cf29-42b8-b17a-a3f95a8e08a3.png)

"Set-Cookie header was blocked due to user prefernces."

Interestingly I haven't changed any of my preferences in Chrome. So I tried Firefox, and while Firefox doesn't report an error, it doesn't set the cookie either. 

![image](https://user-images.githubusercontent.com/4595662/141275995-06a20e51-ad46-4993-8e53-12ffa5f1d88e.png)

EDIT:
At the moment, it's completely breaking the application because when I try to logout it says "refresh_token" is required. Though I am unsure why it is required for logout. 

https://samesite-sandbox.glitch.me/
![image](https://user-images.githubusercontent.com/4595662/141279030-853832ac-3c84-4c49-9ea2-0be6f6fc59d0.png)

Could be something to do with LetsEncrypt. 

EDIT 2:
Renewed all my SSL certificates and it still isn't setting the cookie, I noticed that `SameSite=Lax` is being passed, but it should be `SameSite=None`, https://stackoverflow.com/a/63248851

Interestingly this issue has only just started appearing however. I have no idea why it started but it seems to co-inside with a server side update I had with aaPanel. 

EDIT 3:
In my `.env` file I have changed the refresh token settings to:
```
REFRESH_TOKEN_COOKIE_SECURE=true
REFRESH_TOKEN_COOKIE_SAME_SITE="none"
```
I now have cookies successfully setting, but not in Incognito Mode. Which is expected:
![image](https://user-images.githubusercontent.com/4595662/141283603-e654d4d8-5b1a-47f0-a1b3-389a9e7219cc.png)

I wonder if it's worth setting the defaults to this in future updates? I am also interested why others haven't had this issue, unless they knew this and I'm an idiot (most likely the case). 

EDIT 4:
```
portal.***.dev/auth/refresh:1 Failed to load resource: Failed to load resource: net::ERR_NAME_NOT_RESOLVED
axios-transport.js:130 Uncaught (in promise) Error: Network Error
    at AxiosTransport.<anonymous> (axios-transport.js:130)
    at Generator.throw (<anonymous>)
    at rejected (axios-transport.js:6)
```

![image](https://user-images.githubusercontent.com/4595662/141296797-ee46dc5a-ee55-44ef-9e5b-c91a41bb077e.png)
It's no longer showing the error, but the Cookie isn't being set either: 
![image](https://user-images.githubusercontent.com/4595662/141296886-ae8059ff-20a4-4180-811a-32ba9ba40372.png)

**Reply by rijkvanzanten** [reply]
> "Set-Cookie header was blocked due to user prefernces."
> Interestingly I haven't changed any of my preferences in Chrome. So I tried Firefox, and while Firefox doesn't report an error, it doesn't set the cookie either.

This is most likely the browser preventing cross-origin cookie references by default. Make sure you're using HTTPS both locally and on Directus, and configure [the `REFRESH_COOKIE_` environment variables](https://docs.directus.io/configuration/config-options/#security) accordingly to make sure Directus sets cookies in a way the browser is happy with.

> At the moment, it's completely breaking the application because when I try to logout it says "refresh_token" is required. Though I am unsure why it is required for logout.

What happens on logout is that the session is removed from directus_sessions. The session is tied to the refresh token, as the refresh token allows you to refresh the given session. If the API doesn't know the session, it can't know what session to destroy 👍🏻 

> Renewed all my SSL certificates and it still isn't setting the cookie, I noticed that SameSite=Lax is being passed, but it should be SameSite=None, https://stackoverflow.com/a/63248851

Yeah! That can be controlled through the above mentioned variables 🙂 We're currently opting for Lax by default, as that seems to work for most as of right now.

> In my .env file I have changed the refresh token settings to:

Ah yes, you found them by yourself 😄 (I should read the whole message before starting to write a response, oh well)

> I wonder if it's worth setting the defaults to this in future updates? I am also interested why others haven't had this issue, unless they knew this and I'm an idiot (most likely the case).

Very good question. I believe part of the puzzle is that it needs the `Secure` flag, which in turn requires SSL. If we'd default to that, no localhost installations would run anymore, so that's not really something I'd like to commit to if we can help it 🤔

**Comment by joselcvarela** [comment]
Are you running on `localhost`? Try `127.0.0.1` instead on Chrome and Firefox and check if that works.
I am moving this into a discussion since it seems there's no issue but a misconfiguration.

**Comment by joeinnes** [comment]
So I think this was properly an issue, because it seems if the SDK is loaded in to a page which has a valid auth_token in storage, then the auth_token will be loaded but the timer is never initialised, and so the auth_token will expire and no refresh will take place.

It seems to be a regression introduced here: https://github.com/directus/directus/commit/23f6baa4db019fa4d6093f48b9a251d486257e66#diff-22768c0ce910bc2b5bab168b0008a2d1a05c16dd41675555ea940a8d2b4f45b0 - on line 34 previously there was a token refresh attempt when the SDK constructor ran, which no longer happens.

Good news is that I think adding something like the below into the constructor should fix it (replacing https://github.com/directus/directus/blob/663ba8c23f6755a8fa9f71d78a977642f57c6c27/packages/sdk/src/base/auth.ts#L37).

```
if (this.autoRefresh) {
  try {
    this.autoRefreshJob();
  } catch {
    // Ignore error
  }
} else {
  this.timer = false;
}
```

<details><summary>Diagnostic process</summary>

* using 127.0.0.1 does not seem to allow any cookies to be set
* using localhost allows a cookie to be set - `directus_refresh_token`
* the auth_token is stored in localstorage
* while the auth_token is valid, the Bearer token is properly set on SDK requests, and the `directus_refresh_token cookie` is sent properly
* while the auth_token is not valid, the Bearer token is still set on the SDK request, and the token is rejected as expired.
* I can observe successful posts to `/auth/refresh` on both server and client
* After the refresh posts, localstorage is properly updated.
* OK I think I found the problem - if I refresh the page, then I no longer see any hits to `/auth/refresh`, and the token expires, meaning the next request after the TTL expires will fail.

</details>

<details><summary>Additional Observations I made in the meantime</summary>

- https://github.com/directus/directus/blob/663ba8c23f6755a8fa9f71d78a977642f57c6c27/packages/sdk/src/base/auth.ts#L29 - I can't see where `this.mode` is set (in case no options are passed)
- https://github.com/directus/directus/blob/663ba8c23f6755a8fa9f71d78a977642f57c6c27/packages/sdk/src/base/auth.ts#L14 timeout is 30000 ms by default, but the token coming from directus is apparently valid for 900000 ms
- https://github.com/directus/directus/blob/663ba8c23f6755a8fa9f71d78a977642f57c6c27/packages/sdk/src/base/auth.ts#L66 would evaluate to 900000 - 30000 which explains the above, the 30000 means 'refresh the token 30s before it expires'
- https://github.com/directus/directus/blob/663ba8c23f6755a8fa9f71d78a977642f57c6c27/packages/sdk/src/base/auth.ts#L70 if we fail here then we do nothing? seems odd... why not at least log that the refresh attempt failed, possibly also call `this.logout()`?

</details>

**Reply by TomS-** [reply]
> So I think this was properly an issue, because it seems if the SDK is loaded in to a page which has a valid auth_token in storage, then the auth_token will be loaded but the timer is never initialised, and so the auth_token will expire and no refresh will take place.
> 
> It seems to be a regression introduced here: [23f6baa#diff-22768c0ce910bc2b5bab168b0008a2d1a05c16dd41675555ea940a8d2b4f45b0](https://github.com/directus/directus/commit/23f6baa4db019fa4d6093f48b9a251d486257e66#diff-22768c0ce910bc2b5bab168b0008a2d1a05c16dd41675555ea940a8d2b4f45b0) - on line 34 previously there was a token refresh attempt when the SDK constructor ran, which no longer happens.
> 
> Good news is that I think adding something like the below into the constructor should fix it (replacing
> 
> https://github.com/directus/directus/blob/663ba8c23f6755a8fa9f71d78a977642f57c6c27/packages/sdk/src/base/auth.ts#L37
> 
> ).
> ```
> if (this.autoRefresh) {
>   try {
>     this.autoRefreshJob();
>   } catch {
>     // Ignore error
>   }
> } else {
>   this.timer = false;
> }
> ```
> 
> Diagnostic process
> * using 127.0.0.1 does not seem to allow any cookies to be set
> * using localhost allows a cookie to be set - `directus_refresh_token`
> * the auth_token is stored in localstorage
> * while the auth_token is valid, the Bearer token is properly set on SDK requests, and the `directus_refresh_token cookie` is sent properly
> * while the auth_token is not valid, the Bearer token is still set on the SDK request, and the token is rejected as expired.
> * I can observe successful posts to `/auth/refresh` on both server and client
> * After the refresh posts, localstorage is properly updated.
> * OK I think I found the problem - if I refresh the page, then I no longer see any hits to `/auth/refresh`, and the token expires, meaning the next request after the TTL expires will fail.
> 
> Additional Observations I made in the meantime
> * https://github.com/directus/directus/blob/663ba8c23f6755a8fa9f71d78a977642f57c6c27/packages/sdk/src/base/auth.ts#L29
>    - I can't see where `this.mode` is set (in case no options are passed)
> * https://github.com/directus/directus/blob/663ba8c23f6755a8fa9f71d78a977642f57c6c27/packages/sdk/src/base/auth.ts#L14
>    timeout is 30000 ms by default, but the token coming from directus is apparently valid for 900000 ms
> * https://github.com/directus/directus/blob/663ba8c23f6755a8fa9f71d78a977642f57c6c27/packages/sdk/src/base/auth.ts#L66
>    would evaluate to 900000 - 30000 which explains the above, the 30000 means 'refresh the token 30s before it expires'
> * https://github.com/directus/directus/blob/663ba8c23f6755a8fa9f71d78a977642f57c6c27/packages/sdk/src/base/auth.ts#L70
>    if we fail here then we do nothing? seems odd... why not at least log that the refresh attempt failed, possibly also call `this.logout()`?

Thank you so much for your research into this, we have had continuing issues that seemed to appear out of nowhere. I assumed it may have been a chrome update it appears that it could be further than that.

**Reply by TomS-** [reply]
I'm still being logged out:
![image](https://user-images.githubusercontent.com/4595662/142012624-9aa3d06a-7372-43cb-aa8e-4123538842bb.png)
The cookie is now not reporting any issues but: 
![image](https://user-images.githubusercontent.com/4595662/142012760-268a7664-7dc0-4543-9f57-07818dd37320.png)

No cookies are being set. Both Directus and the front-end are using `https`. I'm totally baffled why it seemly accepts the cookie but doesn't set it.

**Reply by joselcvarela** [reply]
Are you requesting from different domains?
It could be a CORS problem.

**Reply by TomS-** [reply]
@joselcvarela same domain, different sub-domains. No cors errors are being reported from the xHR request. Happy to let someone remote desktop.

**Comment by TomS-** [comment]
This issue is still present and the trouble is, we had a Quiz App go live and the client isn't too happy because revisiting users get stuck logged in but can't do anything. This is because `auth_token` is set, but no `refresh_token` was actioned, so `directus.auth.token` is true but `directus.users.me` can't be read as it's expired and they can not logout because logging out requires a `refresh_token` set in the cookie. So they are stuck in limbo. I also have `Cookie: Secure` and `SameSite=None`.

**Reply by joselcvarela** [reply]
Like joeinnes said, this should be an issue on revisits. 

This PR should solve the issue https://github.com/directus/directus/pull/9777

While this PR is not merged, my suggestion is manually calling refresh like (not tested):

```js
import { Directus } from "@directus/sdk";

const directus = new Directus('')

let timeout = null

async function refreshJob() {
  clearTimeout(timeout)

  try {
    await directus.auth.refresh()

    setTimeout(refreshJob, 10 * 60 * 1000)
  } catch(error) {
    goToLoginPage();
  }
}

refreshJob()
```

**Comment by joeinnes** [comment]
Bringing conversation back to this topic, rather than on the PR.

> Hey guys,

> Thanks for looking into this. It's interesting to me that I didn't have this issue and the cookie would set correctly. While I see this is to do with the refresh not happening. Would this correct the fact that the cookie isn't being set?

Actually, the cookies were being set even in this case. Normally though, browsers will automatically delete expired cookies, so by the time the unwanted behaviour was present, the cookie would have already expired and been deleted. The PR that was merged the other day should fix this.

> So what I'm seeing is:
> 
> set-cookie: directus_refresh_token=<snip>; Max-Age=604800; Path=/; Expires=Wed, 01 Dec 2021 13:04:24 GMT; HttpOnly; Secure; SameSite=None
>
> But under Inspect Element > Application > Storage: Cookies I'm seeing no cookie be set but no errors from the set-cookie header.

So that all looks normal, the header is correct and the expiry is set to a future date (presumably exactly 7 days in the future). It's a good idea to remove any tokens from data you post in public forums, so I've marked it as `<snip>`, but that did look like a valid token. If this is already on a production site, I'd suggest you invalidate all sessions now out of an abundance of precaution (you can change your `SECRET` env var and restart Directus. All currently logged in users will need to log in again before interacting with your app).

> EDIT:
> If I go to the domain Directus is hosted on, I see that the refresh_cookie is set there. I'm assuming that this as intended. I actually haven't looked into cookies with Directus much as they have mostly just worked™ until recently.

Sorry for being blunt but at this point, I'm not sure I understand what the problem is that you're seeing. Did you try the code that @joselcvarela posted above to manually run the refresh? The PR has now been merged, so I'd imagine that the actual fix will be in the next release, but in the meantime, you'll need to manually fetch the new auth_token to reliably maintain authentication between page reloads.

**Reply by rijkvanzanten** [reply]
Didn't realize we were continuing that thread here, so I'll repost my earlier response for clarification:

---

> If I go to the domain Directus is hosted on, I see that the refresh_cookie is set there. I'm assuming that this as intended. I actually haven't looked into cookies with Directus much as they have mostly just worked™ until recently.

@TomS- I'm assuming that's due to the fact that wherever you're logging in from [doesn't allow third-party cookies from being said due to the browsers tracking protection settings](https://developers.google.com/search/blog/2020/01/get-ready-for-new-samesitenone-secure) 🤔 There's [a handful of environment variables](https://docs.directus.io/configuration/config-options/#security) that allow you to control the specific parts of the cookie, like secure and same-site. Try setting `REFRESH_TOKEN_COOKIE_SECURE` to `true` and `REFRESH_TOKEN_COOKIE_SAME_SITE` to `None`

**Reply by TomS-** [reply]
@joeinnes
>so I've marked it as <snip>, but that did look like a valid token. If this is already on a production site,

Thank you for the heads up, this site is currently not in production.

>Sorry for being blunt but at this point, I'm not sure I understand what the problem is that you're seeing. 

No need to apologise for being blunt. That is completely fine with me. I'm just simply asking questions to learn and better my knowledge around the subject.

@rijkvanzanten 
>Try setting REFRESH_TOKEN_COOKIE_SECURE to true and REFRESH_TOKEN_COOKIE_SAME_SITE to None

Those are in place and I see no errors on `set-cookie`. I can not see the cookie (`Under Inspect Element > Application > Cookies`) on the site you login with using the SDK (`www.domain.com`) but I see the cookie set on the Directus install (`directus.domain.com`). I just wondered if that was fine.

**Reply by rijkvanzanten** [reply]
If the cookie doesn't show up in the devtools, it's most likely because the browser (silently) ignored setting it for privacy reasons.. Very helpful when it comes to preventing facebook from reading along with your internet usage, but very very annoying when it comes to developing websites/apps :| 

What exact browser (+ version) are you using?

**Reply by TomS-** [reply]
@rijkvanzanten

I'm using `Chrome Version 96.0.4664.45` (Windows 11) all settings are default.

**Comment by joeinnes** [comment]
Are you developing locally? If so, can you try in Firefox?

Chrome doesn’t consider local host secure and won’t set HttpOnly cookies.

Joe

**Reply by TomS-** [reply]
https://user-images.githubusercontent.com/4595662/143428356-7fbd5691-e4c9-4b23-ae00-10270d3d2b27.mp4

@joeinnes this might answer any questions you have. I'm not developing locally.

FIREFOX:

https://user-images.githubusercontent.com/4595662/143429838-28115d87-4ab2-4c53-a641-b0de04edbb98.mp4

**Reply by joeinnes** [reply]
Thanks, that's interesting.

One thing I see is that the response is coming from portal.wearefyi.dev but the access-control-allow-origin is set to connectv3.wearefyi.dev. I'm assuming that connectv3.wearefyi.dev is where your front end is hosted, and is where you're trying to log into the app?

If so, I don't see a reason that the cookie would not be being set, the cookie string looks valid, and assuming you're connecting over HTTPS, which it seems you are, then everything seems OK.

One thing that does seem clear is that Rijk's instinct appears to be correct, the issue is not that Directus is not behaving properly - Directus is trying to set the proper cookie with the proper access control headers.

To try to narrow down the troubleshooting, can you confirm if this works in a different browser? Do you have any adblockers or other extensions which would block third party cookies? I believe Chrome will allow the SameSite=None with the Secure attribute by default (although this is an option, so might be worth double-checking if that's enabled under Privacy & Security -> Site Settings -> Cookies -> Block third-party cookies and site data, and the Access-Control-Allow-Origin header seems correct.

Right now, I think the next thing we need to do is validate the assumption that this problem is specific to your browser, and then we can work forwards from there.

**Reply by TomS-** [reply]
@joeinnes 

>I'm assuming that connectv3.wearefyi.dev is where your front end is hosted, and is where you're trying to log into the app?

This is correct and both Directus and the app is using `HTTPS` - the two videos one is Chrome and the other is Firefox. Both of which are not setting the cookie. 

>Privacy & Security -> Site Settings -> Cookies -> Block third-party cookies

Is set to `Block third-party cookies in Incognito`. Firefox is set to `Standard`. Next I tried Edge and that is doing the same. No cookie or ad-blockers are in place. 

Is it possible that something is wrong with `NGINX` that is causing this issue?

**Reply by joeinnes** [reply]
Hi,

Sorry, didn't see the Firefox one. On the Firefox video, I see a lot of requests returning 200s for GETs and 204s for OPTIONS, which is what you'd expect to see if the request is being authenticated (not the 401s). I guess that this is because your access token is valid, but I'm not sure why the refresh token is not being set.

My initial hunch was that this was an issue with the browser somehow, but if it's occurring both on Firefox and on Chrome, it seems unlikely that this is the case.

Sorry, I think I'm reaching the end of my ideas with what could be causing this. It certainly seems as though Directus is properly responding to the login and asking the browser to set the cookie, and the browser is not doing so. This header is reaching your browser though, so it seems unlikely that it would be something that Nginx is doing, but as I don't understand why the cookies is not being set, I can't be confident in ruling it out completely.

The only thing I can think of left to try is changing the Access-Control-Allow-Origin to * to see if that works. But even if that works, then I'm afraid I don't know how you would be able to re-secure it afterwards. 

Joe

**Reply by TomS-** [reply]
Hi @joeinnes 

Adding this to my `.env` file:
```
CORS_ORIGIN="*"
```
Then gives me the error
`Access to XMLHttpRequest at 'https://portal.wearefyi.dev/auth/login' from origin 'https://connectv3.wearefyi.dev' has been blocked by CORS policy` 

I'm really really stumped on this one.

**Comment by joeinnes** [comment]
I don’t think it should matter, but what if you remove the quotation marks?

Joe

**Comment by joselcvarela** [comment]
Hello,
I have checked that there's a unique configuration to make it work across multiple domains.
Can you test the following @TomS- ?
```sh
REFRESH_TOKEN_COOKIE_DOMAIN="XXXX" # This should be the domain where Directus is running, i.e. the domain part of PUBLIC_URL
REFRESH_TOKEN_COOKIE_SECURE="true"
REFRESH_TOKEN_COOKIE_SAME_SITE="None"
```

Source: https://github.com/directus/directus/discussions/9788#discussioncomment-1707352

**Reply by TomS-** [reply]
@joselcvarela Nope, still no cookie being set. 
@joeinnes Removing the quotes still gave the CORS issue. 

I'm really stumped on this one, I wonder if it has something to do with the server side of things? I'm using https://www.aapanel.com/index.html with Digital Ocean.

**Reply by TomS-** [reply]
Happy to move hosting to draw a bridge under this. Anyone used Render with Directus?

**Reply by joselcvarela** [reply]
Can you show what is in `REFRESH_TOKEN_COOKIE_DOMAIN`?

**Reply by TomS-** [reply]
`REFRESH_TOKEN_COOKIE_DOMAIN="portal.wearefyi.dev"`

**Reply by joselcvarela** [reply]
Can you confirm if in `Set-Cookie` header includes `Domain`?

Here is my example:
```
Set-Cookie: directus_refresh_token=lVKrVdTtKYkHO4LPIUHSCTBypYs4FsvDIHfiSjINILT47bfFN_SbuL6kq68uRCua; Max-Age=604800; Domain=localhost; Path=/; Expires=Tue, 07 Dec 2021 13:58:21 GMT; HttpOnly; Secure; SameSite=None
```

In my scenario I am accessing a React client in `127.0.0.1:3000` requesting authentication using SSO to my Directus instance on `localhost:8055`. Since one is localhostl and the other is 127.0.0.1, they are considered different domains.

**Reply by rijkvanzanten** [reply]
@joselcvarela I believe browsers might behave differently for localhost vs "real" domains (stupid, I know..). Did you try with "actual" domains? You should be able to reproduce using something like ngrok to make a fake domain tunnel to your localhost.

**Reply by TomS-** [reply]
`set-cookie: directus_refresh_token=vmT793--Ru76hevIwJIasnsxIO5ZojOER_KA74fwA37SfqAXFwefFEWjwe; Max-Age=604800; Domain=portal.wearefyi.dev; Path=/; Expires=Tue, 07 Dec 2021 14:05:16 GMT; HttpOnly; Secure; SameSite=None`

This is what is being returned in `set-cookie`. I have generated some test logins so you can check for yourself.
URL: `https://connectv3.wearefyi.dev/`
Username: `directus@testing.com`
Password: `DirectusTesting!`

**Reply by joselcvarela** [reply]
Humm, it seems to be working @TomS-
<img width="1512" alt="Screenshot 2021-11-30 at 14 54 00" src="https://user-images.githubusercontent.com/14039341/144070338-d43ddd91-5455-4c24-b9eb-239b9ab2acca.png">

**Reply by joeinnes** [reply]
I am also able to log in, but the cookie does not get set. I'm looking at it.

**Reply by joeinnes** [reply]
D'oh... yes, the cookie gets set, but of course it's set on portal.wearefyi.dev, which is why it's not visible in the cookie storage section on connectv3.wearefyi.dev.

Now I checked that, and it looks fine.

After authenticating, the request is going wilth the refresh token attached:

```
GET /users/me?fields[]=customer.fusioo&fields[]=first_name&fields[]=last_name&fields[]=email HTTP/2
Host: portal.wearefyi.dev
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:94.0) Gecko/20100101 Firefox/94.0
Accept: application/json, text/plain, */*
Accept-Language: en-GB,en;q=0.5
Accept-Encoding: gzip, deflate, br
Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6Ijc0NjA0ZDk2LTFiNzAtNDQwMS1hNjcwLWFmNWYzNjEyYTVjMiIsImlhdCI6MTYzODI4NTg5OSwiZXhwIjoxNjM4Mjg2Nzk5LCJpc3MiOiJkaXJlY3R1cyJ9.RFdK7N44Z4N2CRZfzglAFnRCrrd1GD_Ow3ZjvNR1H5c
Origin: https://connectv3.wearefyi.dev
DNT: 1
Connection: keep-alive
Referer: https://connectv3.wearefyi.dev/
Cookie: directus_refresh_token=a3_vMZoXAOV7Reo74cwXynl6Nrz8NQyA-hRpAfz__7TYxNwAjBtngJfzo6Qn_e_q
Sec-Fetch-Dest: empty
Sec-Fetch-Mode: cors
Sec-Fetch-Site: same-site
Pragma: no-cache
Cache-Control: no-cache
```

**Reply by joselcvarela** [reply]
@TomS- 
These are my conclusions:
- Cookie is not being set because you are restrincting the cookie domain to `portal.wearefyi.dev`. Try set "REFRESH_TOKEN_COOKIE_DOMAIN" to `wearefyi.dev`
- Even though, if I close the browser and open it again, I still am logged in. This happens because, even though you can't see the cookie on `Cookies` tab it still exists under `portal.wearefyi.dev` and if you check the requests tab, you can see the cookie is being sent to `portal.wearefyi.dev` and this important part.

In conclusion, I believe your configuration is right and you should not have any issues. Although you can make cookie accessible across your subdomains by setting it up to
```sh
REFRESH_TOKEN_COOKIE_DOMAIN="wearefyi.dev"
```

Let us know your feedback 👍

**Reply by TomS-** [reply]
@joselcvarela 

I see the cookie being set now! :-) 

>Even though, if I close the browser and open it again, I still am logged in.

I found while this happened, after 15 minutes I would get a Token error. With me now seeing the cookie being set I'll let you know how I get on.

Out of interest, what would be the `refresh_token_cookie_domain` be if they were on complete different domains?

EDIT:
![image](https://user-images.githubusercontent.com/4595662/144083971-266dc01d-a5e5-451c-8fb2-b89930f8c01c.png)

Still logging me out, could you please login and refresh after 15 minutes to see if you get logged out?

**Reply by joeinnes** [reply]
`REFRESH_TOKEN_COOKIE_DOMAIN="wearefyi.dev"` just means you'll be able to see the cookie on the storage section of any subdomain of `wearefyi.dev`. I don't know exactly what the issue was for you, it was potentially a blend of the bug we spotted right at the beginning (which should have been fixed in recent releases of the SDK) compounded by a CORS misconfiguration.

At some point, it seems the stars aligned and you fixed it (maybe through fiddling with something, who knows), but I was dragging everyone down the wrong path by being confused that I couldn't see the cookie in Dev Tools, even though that was correct and normal behaviour (as the cookie was being set for a different host). As a result, at some point along the way, the issue got fixed, and we were still running around trying to understand what the problem was when there was no longer any problem.

I believe everything is working fine now.

**Reply by joselcvarela** [reply]
@TomS- 
It will be the domain of your Directus instance. Even though you can't see it on Cookies tab, it is defined on that domain.
Take this example:
1. You set `REFRESH_TOKEN_COOKIE_DOMAIN="wearefyi.dev"` because your Directus instance is running on this domain
2. You have a client on `myserver.com`. In this client you do all the steps required for authenticate with your instance.
3. If you go into `myserver.com` and check Cookies tab, no cookies will be shown, because the cookie is not in this domain.
4. But if you go into `wearefyi.dev` and check Cookies tab, it will show the cookie you requested on `myserver.com`.

This usually happens with Google Analytics and Facebook pixel.

**Reply by TomS-** [reply]
Just in case the edit didn't fire a notification:
![image](https://user-images.githubusercontent.com/4595662/144084457-58cd9fd0-3952-4ff6-a35f-0aaf7c0dc86c.png)
Still logging me out, could you please login and refresh after 15 minutes to see if you get logged out?

EDIT:
![image](https://user-images.githubusercontent.com/4595662/144087353-35c15462-0aae-43f9-94de-313675329e93.png)

Maybe this has something to do with it? It's setting 3 different refresh tokens.

**Reply by joselcvarela** [reply]
I also experienced something alike. In my case was `Invalid User Credentials`.
But I believe it's because of multiple configurations you have tried, now there are some invalid refresh tokens on cookies. Try to clear cookies and check if happens again. I hope not 😅

**Reply by joselcvarela** [reply]
An easy way to clear cookies

https://user-images.githubusercontent.com/14039341/144089562-b67f0db5-7355-49ff-9054-ec78cb3b6862.mov

**Reply by TomS-** [reply]
@joselcvarela 
![image](https://user-images.githubusercontent.com/4595662/144092968-1cc07360-52c3-447c-a111-ebd3d4ef6dfd.png)

Still happening. But this time, if I refresh the page it doesn't actually log me out. Maybe it's something to do with Vite and HMR?

**Reply by TomS-** [reply]
@joeinnes / @joselcvarela This is still happening to me, and this is a built website so can't be to do with Vite and HMR. It prevents logout too, so for clients they just get stuck on this screen. I'm inclined to think that it's the server causing the issue, since I seem to be the only one experiencing it. 

https://user-images.githubusercontent.com/4595662/144578261-4f068ecb-cc3f-4f63-a566-910a3bd72458.mp4
